### ISSUE TITLE
getting very different results when using pinv from jax compared to numpy version

### ISSUE THREAD

**Issue by iamyihwa** [google/jax#16715]
### Description

Hello 
Thanks for the great library ! 
I am using  pinverse from both libraries and the results I get from jax version sometimes is nan, and also is farther away from the result i should be getting. 
Any quick thoughts? 

``` 
import numpy as np 
import jax.numpy as jnp
np.linalg.pinv(M)  
jnp.linalg.pinv(M)
``` 

### Which accelerator(s) are you using?

CPU

### Additional system info

3.8

**Comment by hawkinsp** [comment]
Can you share the input matrix `M`? It's not possible to say without a reproducible example.

**Comment by jakevdp** [comment]
Note that some difference would be expected, because JAX defaults to operating in float32 precision, while numpy defaults to float64. See [JAX Sharp Bits: Double Precision](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#double-64bit-precision) for some info on this.

**Comment by clemisch** [comment]
I have run into this recently with repro below. 

> because JAX defaults to operating in float32 precision

@jakevdp Is this really applicable here? If I set `rcond=1e-15` in the repro below, the result from JAX is correct despite `float32` inaccuracy. 

Maybe the default `rcond` value in JAX could be verified?

---

```python
import numpy as np
import jax.numpy as jnp

np.set_printoptions(linewidth=120, precision=3)

x = np.array(
    [
        -1.5960813392e+02,  +0.0000000000e+00, -2.8550970394e+02, +1.8983549385e+05,
        +3.2236907152e+01,  +3.2469775000e+02, -2.3547361072e+01, -2.1463178817e+03,
        +1.2920604069e-01,  +0.0000000000e+00, -9.4378200689e-02, +3.1609527352e+02,
    ]
).reshape(3, 4)

y = np.linalg.pinv(x)
z = jnp.linalg.pinv(x)
z_ = jnp.linalg.pinv(x, rcond=1e-15)

print(y)
print()
print(z)
print()
print(z_)
```

prints

```
[[-2.178e-03  2.620e-04  1.310e+00]
 [ 6.477e-05  3.033e-03 -1.830e-02]
 [-2.112e-03 -2.716e-04  1.267e+00]
 [ 2.596e-07 -1.882e-07  3.006e-03]]

[[ 3.209e-06  2.842e-04  1.018e-08]
 [ 3.429e-05  3.033e-03  1.087e-07]
 [-2.835e-06 -2.501e-04 -8.976e-09]
 [ 5.266e-06 -1.371e-07  8.767e-09]]

[[-2.178e-03  2.619e-04  1.310e+00]
 [ 6.477e-05  3.033e-03 -1.830e-02]
 [-2.112e-03 -2.716e-04  1.267e+00]
 [ 2.595e-07 -1.883e-07  3.006e-03]]
```

**Comment by jakevdp** [comment]
It looks like numpy and jax use different defaults for `rcond` – in particular, JAX uses an `rcond` based on `np.finfo(dtype).eps`, so it will be different for float32 vs float64 input. NumPy always uses and `rcond=1E-15`, and always casts inputs to float64, regardless of the input type.

**Comment by clemisch** [comment]
Thanks for your timely response Jake.

Yes, the difference is documented. It still took me by surprise and somehow feels different that the typical single vs. double difference, because the repro shows that JAX can achieve high accuracy for smaller `rcond` even with single precision.

Is there a resource/reason for the different `rcond` default?

**Comment by jakevdp** [comment]
It looks like this was established in the very early days of JAX, and was modeled after tensorflow. I found this discussion in the PR that contributed `pinv`: https://github.com/google/jax/pull/1656#discussion_r345020024

**Comment by waterboy96** [comment]
Hi,

I have a small project where I've been trying to replace Numpy with Jax, and I have run into a similar problem. 

I have a feeling that the sparser the matrix becomes, the larger the error is between the Numpy and Jax implementation even when using rcond = 1e-15. However, I am not quite sure how to prove it yet. 

I've created a small example in Colab here: https://colab.research.google.com/drive/17-QWd_RvTFNvM_5QtuEGTNeJ1Yn-TucR?usp=sharing

**Comment by jakevdp** [comment]
Thanks - I think the issue is that your covariance is ill-conditioned (The condition number is $2^{10}$) which means that for operations like the pseudo-inverse, normal floating point roundoff errors will dominate. JAX's `pinv` and numpy's `pinv` have virutally the same implementation, but the results of the operations will not be bitwise-identical as is typical when comparing different approaches to the same floating point computation. When the condition number is high, those floating point rounding errors will begin to dominate the output.

There's nothing really for us to fix here I think – but if your computation is sensitive to the precise numerics of an ill-conditioned pseudo-inverse, it suggests to me that you should look for a different approach to your problem.
